src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -202,17 +211,29 @@ def minimum_python_version(self, interpreter_universe: Iterable[str]) -> str | N
         The constraints may also be compatible with later versions; this is the lowest version that
         still works.
         """
+        snapped = self.snap_to_minimum(interpreter_universe)
+        if snapped:
+            # We know by construction that snapped has a single requirement, and that has a single
+            # (op, version) spec whose version is "major.minor.*", so we just strip off the ".*" .
+            return next(iter(snapped)).specs[0][1][:-2]
+        return None
+
+    def snap_to_minimum(self, interpreter_universe: Iterable[str]) -> InterpreterConstraints | None:
+        """Snap to the lowest Python major.minor version that works with these constraints."""
         for major, minor in sorted(_major_minor_to_int(s) for s in interpreter_universe):
-            if self._includes_version(major, minor):
-                return f"{major}.{minor}"
+            for p in range(0, _PATCH_VERSION_UPPER_BOUND + 1):
+                for req in self:
+                    if req.specifier.contains(f"{major}.{minor}.{p}"):  # type: ignore[attr-defined]
+                        snapped = Requirement.parse(f"{req.project_name}=={major}.{minor}.*")
```

Comment:
This needs to handle patch versions. That allows users to ignore problematic versions like `2.7.2`, while still using 2.7 generally.

You can use `_valid_patch_versions` to get back all patch versions supported. Here's some code I wrote to then create the requirement string based on that: https://github.com/pantsbuild/pants/pull/12543

Reply:
This replicates the logic of _valid_patch_versions, no?

Reply:
Whether it needs to ignore problematic versions depends on how it will be used, for now though this method does exactly what it says it does (and is equivalent to the existing minimum_python_version() which also did not handle excluding patch versions).

Reply:
No, `_valid_patch_versions` is a helper to give back a sequence of patch versions like `[1, 2, 4, 5]`. This function is instead creating a valid `pkg_resources.Requirement`. You'd be calling `_valid_patch_versions` to convert that example into something like `CPython>=2.7.1,<=2.5,!2.7.3`

Reply:
> does exactly what it says it does (and is equivalent to the existing minimum_python_version() which also did not handle excluding patch versions).

`minimum_python_version` intentionally gives up on patch versions, whereas this function is claiming the patch `.*` is valid. Those are different things. `minimum_python_version` can only be used in contexts that don't care about patch versions, like here:

https://github.com/pantsbuild/pants/blob/b24d1fbc6a8279fb494f3f344a05fd5443653d59/src/python/pants/backend/python/typecheck/mypy/subsystem.py#L218-L225

Reply:
Makes sense, fixed.